Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -1,443 +1,381 @@
-    }
+// Private helper that sets _dispatchQueue based on the current quality of service
+- (void)_setDispatchQueueUsingQualityOfService {
+    std::lock_guard<std::recursive_mutex> lock(_dispatchQueueLock);
+    _dispatchQueue = dispatch_get_global_queue(_QOSClassForNSQualityOfService(self.qualityOfService), 0);
+}
 
-    if (op != nil) {
-        if ([op isReady]) {
-            *curOperation = op;
-            [op start];
-            *curOperation = nil;
-        } else {
-            NSAtomicListInsert(sourceListPtr, op);
-            // We claim that we have done some work here.
-            // Otherwise, this work session is considered to be over and we clear the pending queue.
-            return TRUE;
-        }
-    }
++ (BOOL)automaticallyNotifiesObserversOfOperations {
+    // This class manually notifies for operations, which generates changes in addOperation/s.
+    return NO;
+}
 
-    return op != nil;
++ (BOOL)automaticallyNotifiesObserversOfOperationCount {
+    // This class manually notifies for operationCount, as it changes at the same time as operations
+    return NO;
 }
 
-- (id)_workThread {
-    BOOL stop = FALSE;
+- (instancetype)init {
+    if (self = [super init]) {
+        _operations.attach([NSMutableArray new]);
+        _maxConcurrentOperationCount = NSOperationQueueDefaultMaxConcurrentOperationCount;
+        _qualityOfService = NSQualityOfServiceDefault;
 
-    while (!stop) {
-        [self _doMainWork];
-        pthread_mutex_lock(&priv->_threadRunningLock);
-        if (![self hasMoreWork]) {
-            stop = TRUE;
-            priv->_thread = nil;
+        if (dispatch_queue_t underlyingQueue = [self underlyingQueue]) {
+            _dispatchQueue = underlyingQueue;
+        } else {
+            [self _setDispatchQueueUsingQualityOfService];
         }
-        pthread_mutex_unlock(&priv->_threadRunningLock);
     }
-    [priv->allWorkDone signal];
-
     return self;
 }
 
-- (NSOperationQueue*)_doMainWork {
-    memset(priv->myQueues, 0, sizeof(priv->myQueues));
+- (void)dealloc {
+    [_name release];
 
-    BOOL didWork;
+    for (NSOperation* op in _operations.get()) {
+        [op removeObserver:self forKeyPath:@"isFinished" context:nil];
+        [op removeObserver:self forKeyPath:@"isReady" context:nil];
+    }
 
-    id innerPool = [[NSAutoreleasePool alloc] init];
+    if (_underlyingQueue) {
+        dispatch_release(_underlyingQueue);
+    }
 
-    do {
-        didWork = FALSE;
+    [super dealloc];
+}
 
-        for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
-            [priv->suspendedCondition lock];
-            while (priv->isSuspended) {
-                [priv->suspendedCondition wait];
+- (void)observeValueForKeyPath:(NSString*)keyPath ofObject:(id)object change:(NSDictionary*)change context:(void*)context {
+    if ([keyPath isEqualToString:@"isFinished"]) {
+        NSOperation* op = static_cast<NSOperation*>(object);
+        if (op.isFinished) {
```

Comment:
nit: if we request the 'new' value from the change, we can avoid the synchronized call back into operation to query it.